Start Proxy in TLS only mode.

jai1 opened the pull request:
Two main goals of this PR:-
- Start Pulsar services (broker, proxy, websocket, discovery) in TLS only mode, such that they only listen on TLS ports. 
- Once ServiceUrlTls is set enableTls becomes redundant information - hence getting rid of the flag in relevant components.

merlimat commented on conf/proxy.conf:
```diff
@@ -18,10 +18,10 @@
 
 # Zookeeper quorum connection string (comma-separated)
-zookeeperServers=
+zookeeperServers=localhost:2181
```
Leave empty so that we get an error if not filled out instead of having the zk client trying and retrying to connect

merlimat commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConfiguration.java:
```diff
@@ -41,11 +41,13 @@
     private String brokerServiceURLTLS;
     
     // Port to use to server binary-proto request
+    // Set to -1 to disable the port.
```
Since we have already `enableTLS`, I'd prefer to maintain consistency here and have another separate flag to disable non encrypted port.

merlimat replied:
Also, for the same consistency reason, we should have the same behavior in broker config

jai1 replied:
I was thinking of removing enableTLS, set default value of tls port to -1 
If user want TLS he will explicitly set the value on the tls port 

Why have two extra flags (enableTls and disableNonEncryptProtocol)?

merlimat replied:
I'd prefer to leave it empty then, rather than -1, but have the same behavior for non-TLS port. If it's empty, than it's disabled.

jai1 commented:
@merlimat - have addressed your review comments
Had to change the ports data type to integer else you get 
```
java.lang.IllegalArgumentException: failed to initialize servicePort field while setting value 
	at org.apache.pulsar.common.util.FieldParser.lambda$update$0(FieldParser.java:126)
	at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
	at org.apache.pulsar.common.util.FieldParser.update(FieldParser.java:120)
	at org.apache.pulsar.common.configuration.PulsarConfigurationLoader.create(PulsarConfigurationLoader.java:89)
	at org.apache.pulsar.common.configuration.PulsarConfigurationLoader.create(PulsarConfigurationLoader.java:73)
	at org.apache.pulsar.common.configuration.PulsarConfigurationLoader.create(PulsarConfigurationLoader.java:54)
	at org.apache.pulsar.proxy.server.ProxyServiceStarter.<init>(ProxyServiceStarter.java:79)
	at org.apache.pulsar.proxy.server.ProxyServiceStarter.main(ProxyServiceStarter.java:127)
Caused by: java.lang.IllegalArgumentException: Can not set int field org.apache.pulsar.proxy.server.ProxyConfiguration.servicePort to null value
	at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:167)
	at sun.reflect.UnsafeFieldAccessorImpl.throwSetIllegalArgumentException(UnsafeFieldAccessorImpl.java:171)
	at sun.reflect.UnsafeIntegerFieldAccessorImpl.set(UnsafeIntegerFieldAccessorImpl.java:80)
	at java.lang.reflect.Field.set(Field.java:764)
	at org.apache.pulsar.common.util.FieldParser.lambda$update$0(FieldParser.java:124)
```

I will make changes to the broker side and Discovery Configuration as a part of  a separate PR

merlimat commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConfiguration.java:
```diff
@@ -41,16 +41,18 @@
-    // Set to -1 to disable the port.
-    private int servicePort = 6650;
+    // Set to null to disable the port.
+    private Integer servicePort;
```
By default we should still listen on non-tls port

merlimat commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConfiguration.java:
```diff
@@ -41,16 +41,18 @@
     // Port to use to server binary-proto-tls request
-    private int servicePortTls = 6651;
+    // Set to null to disable the port.
+    private Integer servicePortTls = 6651;
```
If TLS is disabled by default, we should set this defaulting to null, no?

merlimat commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConfiguration.java:
```diff
@@ -146,46 +145,38 @@ public void setZookeeperSessionTimeoutMs(int zookeeperSessionTimeoutMs) {
     }
 
-    public int getServicePort() {
+    public Integer getServicePort() {
```
I prefer to use `Optional<Integer>` here to force the caller to explicitly handle the fact that this can be null.

merlimat commented on pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConfiguration.java:
```diff
@@ -146,46 +145,38 @@ public void setZookeeperSessionTimeoutMs(int zookeeperSessionTimeoutMs) {
         return servicePort;
     }
 
     public void setServicePort(int servicePort) {
```
If this is taking `int`, there's no way to set it to `null`

maskit commented:
I’m also fine with using ‘null’ as “disabled” but how are we going to deal with default values/behavior?

I didn’t read the details but my expectation was:
-1: Disabled
null: Default value (can be -1)
0: Invalid
>0: Enabled with the specified port number

jai1 commented:
@merlimat - have addressed all your comments

jai1 commented on conf/broker.conf:
```diff
@@ -29,13 +29,13 @@ globalZookeeperServers=
 
 # Broker data port for TLS
-brokerServicePortTls=6651
+brokerServicePortTls=
```
Since by default TLS is diabled

maskit replied:
Just to confirm, you mean TLS is disabled by a preset value(null) and non-TLS port will be disabled if I set `brokerServicePort` to null, right?

sijie commented:
@jai1 are we going to include this in 2.0? shall we move it to 2.1?

sijie commented:
@jai1 I moved this to 2.1, if you think we should include this for 2.0, please move it back :)
